mlc_llm/relax_model/llama.py

```diff
@@ -460,14 +456,21 @@ def forward(
             attention_mask=attention_mask,
             all_seq_len_shape=all_seq_len_shape,
         )
+        if self.self_attn.num_shards > 1:
+            residual = nn.emit(residual / R.const(self.self_attn.num_shards, dtype="float16"))
         hidden_states = nn.emit(residual + hidden_states)
-
+        if self.self_attn.num_shards > 1:
+            hidden_states = nn.emit(ccl.allreduce(hidden_states, "sum"))
+            
         # Fully Connected
         residual = hidden_states
         hidden_states = self.post_attention_layernorm(hidden_states)
         hidden_states = self.mlp(hidden_states)
+        if self.mlp.num_shards > 1:
+            residual = nn.emit(residual / R.const(self.mlp.num_shards, dtype="float16"))
```

Comment:
why are you assuming it's "float16"?

Reply:
good point, we should use the dtype of the model